### ISSUE TITLE
Run swmmr on AWS cloud (linux based)

### ISSUE THREAD

**Issue by HamedGhodsi90** [dleutnant/swmmr#47]
Hi dleutnant,

I have developed my R code using swmmr package on my local computer and it works properly. Now, I am moving all my code in AWS (amazon web services) to run my code in the cloud. AWS uses Linux.

As you may know, I had some problems in using swmm5 executive file, so I build it for Linux. Now I am able to run the simulation model using run_swmm() function and get both .rpt and .out files for further analysis. 
The link for swmm5 executive file for Linux:
https://github.com/HamedGhodsi90/CSO_Real-Time_Prediction/blob/master/swmm5

When I use read_out() function in my local computer, it works properly, for example:
read_out(file = "C:/Users/.../SWMM_online.out", iType = 1, object_name = "10058", vIndex = 4)
The result is:
2020-06-22 11:09:00 0.00
2020-06-22 11:24:00 0.45
2020-06-22 11:39:00 0.00
2020-06-22 11:54:00 0.00
2020-06-22 12:09:00 0.29
2020-06-22 12:24:00 0.00
2020-06-22 12:39:00 0.00
2020-06-22 12:54:00 0.00

However, when I use read_out() function in AWS Rstudio cloud, I get an error (attached figure). for example:
read_out(file = "SWMM_online.out", iType = 1, object_name = "10058", vIndex = 4)

Do you know what is the problem? why does this function works on my local computer, but it does not work in the cloud? 
I also tried read_rpt() function, it works properly in the cloud. It seems that the problem is just for read_out() function. 

You can find the .out file here:
https://github.com/HamedGhodsi90/CSO_Real-Time_Prediction/blob/master/SWMM_online.out

I believe if you use it on your local computer, it should work, but to see the problem, I think you should try it in a cloud. 

![Capture](https://user-images.githubusercontent.com/43046511/85333613-e142a200-b4a7-11ea-8c02-9725331feac4.JPG)

**Comment by dleutnant** [comment]
The message „SWMM executable not found“ is raised on package load and indicates that {swwmr} does not find the swmm5 executable at the expected path. Expected paths for non windows os are "/usr/local/bin/swmm5" or "/usr/bin/swmm5".

What is the path to your swmm5 executable? Did you try the `exec` argument from `run_swmm`?

**Comment by HamedGhodsi90** [comment]
I have put my R code and the exe file in the same folder in the cloud. 
For running the simulation model, I used run_swmm and define the exec as the executive file I have in the folder, and it works properly and gives me two .out and .rpt results. But in the next step, when I wanna see the time series of nodes or links in the .out file, I get the error I explained above. 

The path to my folder in AWS cloud that include all needed files is: 
path_SWMM_exe_file  <- "/home/rstudio/OnlineMonitoringV1.2/swmm5"

The location "/home/rstudio/" is the default one when I use Rstudio from AWS EC2 engine. 

So the code for running simulation model is:
run_swmm ("SWMM_online.inp", rpt = "SWMM_online.rpt", out = "SWMM_online.out", exec = path_SWMM_exe_file , stdout = "", wait = TRUE )
This code works and I get these two files (SWMM_online.rpt and SWMM_online.out).

But when I use this code to read the .out file:
aa <- read_out(file = "SWMM_online.out", iType = 1, object_name = "10058", vIndex = 4)
I get the above error. 

Do you know how can I fix it? why in read_out it wants to use the executive file?

**Comment by dleutnant** [comment]
What's the content of `aa` after `aa <- read_out(file = "SWMM_online.out", iType = 1, object_name = "10058", vIndex = 4)`.

**Comment by HamedGhodsi90** [comment]
Nothing. I get this error:
![Capture](https://user-images.githubusercontent.com/43046511/85368026-6f467900-b4f8-11ea-9dcd-42424f7e2080.JPG)

**Comment by HamedGhodsi90** [comment]
Based on your suggestion, I found directory usr/bin in the cloud, and I copy and paste the swmm5 to this directory, and then I run the read_out() function. I get this error. As you can see, "SWMM executable not found" is not in the error anymore. But still, I have the error about masking some packages. 
Do you know how I can fix this problem?

![Capture](https://user-images.githubusercontent.com/43046511/85369413-11676080-b4fb-11ea-9103-5c399dfdce8e.JPG)

**Comment by dleutnant** [comment]
mmh, it's hard for me to track the problem down. Could you please start a fresh R session and execute the following reproducible example? Please change the path to your executable...

``` r
library(swmmr)
# use a known example
inp_file <- system.file("extdata", "Example1.inp", package = "swmmr", mustWork = TRUE)

# create dummy file
tmp_file <- tempfile()
rpt <- paste0(tmp_file, ".rpt")
out <-  paste0(tmp_file, ".out")

# run simulation
run_swmm(inp_file, rpt = rpt, out = out, exec = "/usr/local/bin/swmm5")
#> Argumente 'minimized' und 'invisible' sind nur für Windows

# import the data
results <- read_out(file = out, 
                    iType = 1, 
                    object_name = "9", 
                    vIndex = c(0,4,5))

head(do.call(cbind, results$`9`))
#> Warning: timezone of object (GMT) is different than current timezone ().
#>                     water_depth total_inflow surface_flooding
#> 1998-01-01 01:00:00   0.0000000    0.0000000                0
#> 1998-01-01 02:00:00   0.2887744    1.2396327                0
#> 1998-01-01 03:00:00   0.4184525    2.5954294                0
#> 1998-01-01 04:00:00   0.5676332    4.6454787                0
#> 1998-01-01 05:00:00   0.4250466    2.6753681                0
#> 1998-01-01 06:00:00   0.2389560    0.8425002                0
```

<sup>Created on 2020-06-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by dleutnant** [comment]
The message "SWMM executable not found" is not an error. It rather warns the user that {swmmr} does not find the executable at its predefined path. Reading .out files still works, though. Please check the manual of the function `run_swmm()`.

**Comment by HamedGhodsi90** [comment]
I run your code. 
Again, I get the same error (attached figure). 
It seems that something is wrong when I call read_out() function. because run_swmm() works. but read_out() doe not work.

![Capture](https://user-images.githubusercontent.com/43046511/85370376-9a32cc00-b4fc-11ea-9162-04ebfb5ebd87.JPG)

**Comment by dleutnant** [comment]
Interesting 🤔. Did you attach {DT} and {dplyr} by yourself?

**Comment by HamedGhodsi90** [comment]
The first time I wanted to use Rstudio in AWS, I manually added some packages I needed like shiny, ggplo2, leaflet, dplyr, xts, zoo, ...
I search a little bit, it seems that these are some warning, not an error. but when I run the model I get a message that says the R crashed (below figure), then those warnings. 

![Capture2](https://user-images.githubusercontent.com/43046511/85418969-7e9ce500-b53f-11ea-8011-5abee19aa188.JPG)

It seems there are some problems regarding Rstudio in AWS. Because I do not get this error in my local computer.

**Comment by dleutnant** [comment]
Do you use the same compiler for SWMM5 and {swmmr}? I remember a similar story where RStudio crashed when I used different ones (clang and gcc) on my local machine...  harmonizing my toolchain solved that issue, but not sure whether it applies to your case as well. 

Do you have access to a fresh linux machine (e.g. docker?!) to try your setup before you move over to AWS?

**Comment by HamedGhodsi90** [comment]
Hi dleutnant,
I use this code to test and my AWS RStudio crash. 
```
library(swmmr)
download.file("https://github.com/HamedGhodsi90/CSO_Real-Time_Prediction/raw/master/SWMM_online.out", "SWMM_online.out")
results <- swmmr::read_out(file = "SWMM_online.out", iType = 1, object_name = "10058", vIndex = 4)
```
So I debug it to find the source of the problem. I reach to function "OpenSwmmOutFile"

function (outFile) 
{
	.Call("_swmmr_OpenSwmmOutFile", PACKAGE = "swmmr", outFile)
}

It seems that this function is not in the package or somehow it won't be called or worked. 
Do you have any idea why it happens? all other functions in the package works properly, but function read_out() won't work because, it cannot find OpenSwmmOutFile function.

**Comment by HamedGhodsi90** [comment]
Regarding your question, swmm5 executive file uses cc compiler, and the installed swmmr package in AWS also used cc compiler. 
No, I do not have an access to a fresh Linux server.

**Comment by dleutnant** [comment]
Did you install {swmmr} using `install.packages("swmmr", type = "source")`?

**Comment by HamedGhodsi90** [comment]
Yes I did install swmmr package using install.packages("swmmr", type = "source").
I found your reply to the same question in issue #16 which is:

"""""""
OK, I can reproduce the error. It seems that the issue is related to g++ (or at least with its default compiler flags...). By using clang++ the .out file is read without overflows. Here's what I did:

install clang and its dependencies: sudo apt-get install clang
create Makevars file (more info here) in the directory ~/.R
edit Makevars sudo vi ~/.R/Makevars with the following content:
CC=clang -Qunused-arguments
CXX=clang++ -Qunused-arguments
reinstall swmmr: install.packages("swmmr", type = "source"). You can now check if clang is used. The output should clearly indicate "clang++" :
* installing *source* package ‘swmmr’ ...
** package ‘swmmr’ successfully unpacked and MD5 sums checked
** libs
clang++ -Qunused-arguments -I/usr/share/R/include -DNDEBUG   -I"/usr/local/lib/R/site-library/Rcpp/include"   -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-3.3.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o
clang++ -Qunused-arguments -I/usr/share/R/include -DNDEBUG   -I"/usr/local/lib/R/site-library/Rcpp/include"   -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-3.3.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c swmm5_iface.cpp -o swmm5_iface.o
reproduce the section Initiate a SWMM run and retrieve simulation results from here (hint: don't forget to set the correct path to your freshly compiled SWMM5 executable)

(optional) Delete Makevars file as it could have an impact on other package builds (s. here)

The buffer overflow error should be gone now. Could you please confirm?
"""""

Should I follow these steps to solve my problem? If yes, can you explain more how can I create Makevars file (number 2 in the instruction)? I could not understand it even when I read the provided link.

**Comment by dleutnant** [comment]
Oh yeah!

You need to create a file named "Makevars" in the directory `~/.R` which only has the following text in it:

```
CC=clang -Qunused-arguments
CXX=clang++ -Qunused-arguments
```

if you are on a terminal you could create the file with `sudo touch ~/.R/Makevars`.

**Comment by dleutnant** [comment]
and afterwards you need to add the the lines above using  `sudo vi  ~/.R/Makevars` .

**Comment by dleutnant** [comment]
Btw: Did you update all your packages beforehand?

**Comment by HamedGhodsi90** [comment]
Sorry I am asking some simple questions. I am not too familiar with Linux.
So, I have updated all packages.
Then I installed clang and its dependencies. 
Then I create and edit Makevars, and installed 'swmmr' package. But everything is the same as before like this:
``` 
* installing *source* package ‘swmmr’ ...
** package ‘swmmr’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
g++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG  -I"/usr/local/lib/R/site-library/Rcpp/include"   -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-VAQCff/r-base-3.6.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c RcppExports.cpp -o RcppExports.o
g++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG  -I"/usr/local/lib/R/site-library/Rcpp/include"   -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-VAQCff/r-base-3.6.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c swmm5_iface.cpp -o swmm5_iface.o
g++ -std=gnu++11 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o swmmr.so RcppExports.o swmm5_iface.o -L/usr/lib/R/lib -lR
installing to /home/rstudio/R/x86_64-pc-linux-gnu-library/3.6/00LOCK-swmmr/00new/swmmr/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (swmmr)

The downloaded source packages are in
	‘/tmp/RtmpTq1E3t/downloaded_packages’ 
```

I think I do not create the Makevars file in a proper directory. You said I create it in ~/.R/Makevars directory, but I do not have this directory (~/.R).
Can you tell me where should I make this file? 
I am using the terminal in AWS RStudio. 
the directories are like this:
the R packages will be installed here:    /home/rstudio/R/x86_64-pc-linux-gnu-library/3.6/
I also have these directories: /usr/local/lib/R/ .....

I do not know where should I create this file. Actually I created it and copy and paste it in all directories that I thought it may work. But finally, it did not.

**Comment by dleutnant** [comment]
Ok, you could do this from R as well:

`cat("CC=clang -Qunused-arguments","CXX=clang++ -Qunused-arguments",file="~/.R/Makevars",sep="\n",append=TRUE)'

Then restart your R session and try `install.packages("swmmr")` again.

**Comment by dleutnant** [comment]
Any news?

**Comment by HamedGhodsi90** [comment]
Yes. I could fix it. The problem was that there was no ".R" file directory in my AWS cloud. So, when I tried to create the Makevars file in this directory, I got error that said there is not such a directory. 
To solve it, I just create a ".R" file by myself, then create the Makevars file in this directory and followed your steps (upgrade all packages, restart R section, uninstall previous swmmr package, ...), and finally it worked, which means that I could reinstall the swmmr package with the same compiler (clang). Then I run my code and everything was fine, especially the ```read_out()``` function. 
Thank you for your help to solve this problem. 
I think anybody who wanna use swmmr package in AWS cloud would face the same problem. So, I suggest to put a guideline somewhere in your GitHub to explain how users should act to successfully run the swmmr package in AWS RStudio. 
Do not forget to include creating .R folder, if there is not any! :D
